Starting Typescript Migration

herbertli opened the pull request:
- Added typescript packages to package.json
- Migrated styles.js (changed ext.)
- Migrated firebase.js (changed ext.)
- Migrated constants.js (changed ext.)
- Migrated GridListCustom.js
- Migrated SpecTile.js
- Migrated SpecList.js
- Migrated Piece.js
- Migrated PieceList.js
- Migrated ItemTypes.js
- Migrated ElementTile.js
- Migrated ElementList.js
- Migrated BoardTile.js
- Migrated BoardList.js
- Migrated Logout.test.js
- Migrated Logout.js
- Migrated NavigationMaster.js
- Migrated Sidebar.js
- Migrated Board.js
- Migrated CanvasImage.js
- Migrated Element.js
- Migrated SpecInfo.js
- Migrated SpecInfo.test.js

SpecViewer still needs migrating, but react-contenteditable has no typescript support

@yoav-zibin @apsdehal Sorry, I couldn't figure out how to add a reviewer.
Please let me know if I need to change anything!

apsdehal committed:
Show only user uploaded content everywhere

herbertli committed:
Migrated BoardList, ElementList, PieceList, SpecList

Also modified supporting files when necessary. Full list and details:

- Added typescript packages to package.json
- Migrated styles.js (changed ext.)
- Migrated firebase.js (changed ext.)
- Migrated constants.js (changed ext.)
- Migrated GridListCustom.js
- Migrated SpecTile.js
- Migrated SpecList.js
- Migrated Piece.js
- Migrated PieceList.js
- Migrated ItemTypes.js
- Migrated ElementTile.js
- Migrated ElementList.js
- Migrated BoardTile.js
- Migrated BoardList.js

herbertli committed:
Migrated Navigation Components

- Migrated Logout.test.js
- Migrated Logout.js
- Migrated NavigationMaster.js
- Migrated Sidebar.js

herbertli committed:
Finished Migrating most of gamespec

- Migrated Board.js
- Migrated CanvasImage.js
- Migrated Element.js
- Migrated SpecInfo.js
- Migrated SpecInfo.test.js

SpecViewer still needs migrating, but react-contenteditable has no typescript support

apsdehal commented on src/components/GridListCustom.tsx:
```diff
@@ -1,16 +1,17 @@
-function GridListCustom(Component) {
-  return props => {
+function GridListCustom(Component: React.StatelessComponent | React.ComponentClass) {
+  return (props: any) => {
```
Maybe you can move away from any here also.

apsdehal commented on src/components/gamespec/Board.tsx:
```diff
@@ -274,7 +306,7 @@ class Board extends React.Component {
                     this.handleClickOn(index);
                   }}
-                  item={item}
+                  // item={item}
```
This should not be commented I think?

herbertli replied:
Just wanted to clarify... is item={item} passing down all other props down to the child component, or is there an actual prop called "item"? Thanks!

apsdehal replied:
It is getting passed down, it is used somewhere, but I don't remember where.

apsdehal replied:
Got it, it is used by react-dnd when we drag and drop.

apsdehal commented on src/components/gamespec/BoardTile.tsx:
```diff
@@ -1,10 +1,17 @@
+  selectedKey: any;
+  handleGridTileClick: (keyProp: any, image: any) => void;
+  image: any;
+}
```
Using any doesn't provide any benefits of TypeScript. If possible, can you move away from any?

apsdehal reviewed (changes requested):
Overall, looks good. With some `any` types which can be possibly removed.

Can you do one more thing, if possible can you rebase ts-migrate branch with current master as there are some changes I made there.

apsdehal commented:
Also, try not to push big PRs. It is hard to review and harder to rebase in case some change happens in master. Use small commits, so that they are easy to rebase with new upstream changes. Now, it will be a lot of pain and effort to rebase with master.

herbertli committed:
Merge branch 'master' into ts-migrate

Pull changes from master

herbertli commented:
@apsdehal I think I've managed to get my branch in sync with master, is there a way to test the site locally?

apsdehal commented:
Can you follow the instructions in Readme for setup? Once you have everything I can provide you credentials for Firebase.

apsdehal commented:
@herbertli Were you able to run it?

herbertli commented:
@apsdehal it compiles correctly, just need authentication credentials!

apsdehal commented:
Check piazza

herbertli committed:
Various Bug Fixes and Style Changes

- Downgraded @types/material-ui to be compatible with project (in the future, I think this package will need to be updated)
- Fixed various linting errors
- Changed various onClick handlers (material-ui v.20) back to the old onTouchTap handlers (material-ui v.18.x)

Project now compiles and works (provided you change the "no-any" property in .tslint to false)

herbertli committed:
Bug Fixes

Bunch of bug fixes to make the app compile and run for typescript (if you change the no-any option in tslint to false)

- Downgrade @types/material-ui to correct version
- Null checks for ElementCreatorContainer
- Pass item prop in Board
- Clean up CanvasImage
- Fixes to NavigationMaster to make it work with BrowserRouter
- Updated types in SidebarProps

herbertli committed:
Commit tslint

- Change no-any to false in order to make project compile

apsdehal commented:
@herbertli Consider this complete (make sure there aren't any files left if you are continuing with this homework). I will need some time to pull down rebase and then push stuff.

Thanks.

apsdehal commented:
Ok, I have merged this into master. You can delete your ts-migrate branch and create a new one from master.
